src/pv.rs

```diff
@@ -43,28 +44,118 @@ fn packet_dump(packet: &PvPacket) {
             private: std::ptr::null_mut(),
         }
     }
+
+    /* check whether the new header can be attached to payload from the start of packet
+    add header to payload and return Ok if it's possible, or return Err */
+    pub fn attach_header(&mut self, header: Vec<u8>) -> Result<(), ()> {
+        if self.start >= header.len() as u32 {
+            // header can be attached in empty space of headroom
+
+            self.start = self.start - header.len() as u32;
+
+            unsafe {
+                copy_nonoverlapping(
+                    header.as_ptr(),
+                    self.buffer.offset(self.start as isize),
+                    header.len(),
+                );
+            }
+
+            Ok(())
+        } else if header.len() as u32 <= self.buffer_size - self.end + self.start {
+            // header can't be attached in headroom but can be by shifting payload
+            unsafe {
+                // move payload (memmove)
+                copy(
+                    self.buffer.offset(self.start as isize),
+                    self.buffer.offset(header.len() as isize),
+                    (self.end - self.start) as usize,
+                );
+
+                // set start and end of data
+                self.start = 0;
+                self.end = self.end + (header.len() as u32 - self.start);
+
+                // copy header data (memcpy)
+                copy_nonoverlapping(
+                    header.as_ptr(),
+                    self.buffer.offset(self.start as isize),
+                    header.len(),
+                );
+            }
+
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+
+    // replace payload with new data from *start and make the rest of original payload zero
+    pub fn replace_data_from_start(&mut self, new_data: &Vec<u8>) -> Result<(), ()> {
+        let replaceable_len = self.buffer_size - self.start;
+        let payload_len = self.end - self.start;
+
+        if replaceable_len >= new_data.len() as u32 {
+            // make the rest of original payload zero if new data is smaller than the original payload
+            if payload_len >= new_data.len() as u32 {
+                for i in (self.start + new_data.len() as u32)..self.end {
+                    unsafe {
+                        write(self.buffer.offset(i as isize), 0);
+                    }
+                }
+            }
+
+            self.end = self.start + new_data.len() as u32;
+
+            unsafe {
+                // replace data
+                copy(
+                    new_data.as_ptr(),
+                    self.buffer.offset(self.start as isize),
+                    new_data.len(),
+                );
+            }
+
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+
+    // replace payload with new data from *start and make the rest of original payload zero
+    pub fn get_buffer_as_vec(&self) -> Vec<u8> {
```

Comment:
Just `buffer` is enough for function name

Reply:
Changed it